node/core/pvf/src/executor_intf.rs

```diff
@@ -97,17 +100,58 @@ pub fn prevalidate(code: &[u8]) -> Result<RuntimeBlob, sc_executor_common::error
 
 /// Runs preparation on the given runtime blob. If successful, it returns a serialized compiled
 /// artifact which can then be used to pass into `Executor::execute` after writing it to the disk.
-pub fn prepare(blob: RuntimeBlob) -> Result<Vec<u8>, sc_executor_common::error::WasmError> {
-	sc_executor_wasmtime::prepare_runtime_artifact(blob, &CONFIG.semantics)
+pub fn prepare(
+	blob: RuntimeBlob,
+	executor_params: ExecutorParams,
+) -> Result<Vec<u8>, sc_executor_common::error::WasmError> {
+	let semantics = params_to_wasmtime_semantics(executor_params)
+		.map_err(|e| sc_executor_common::error::WasmError::Other(e))?;
+	sc_executor_wasmtime::prepare_runtime_artifact(blob, &semantics)
+}
+
+fn params_to_wasmtime_semantics(par: ExecutorParams) -> Result<Semantics, String> {
+	let mut sem = DEFAULT_CONFIG.semantics.clone();
+	let mut stack_limit = if let Some(stack_limit) = sem.deterministic_stack_limit.clone() {
+		stack_limit
+	} else {
+		return Err("No default stack limit set".to_owned())
+	};
+	for p in par.iter() {
+		match p {
+			ExecutorParam::Environment(env) =>
+				if *env != ExecutionEnvironment::WasmtimeGeneric {
+					return Err("Wrong execution environment type".to_owned())
+				},
+			ExecutorParam::ExtraHeapPages(ehp) => sem.extra_heap_pages = *ehp,
+			ExecutorParam::MaxMemorySize(mms) => sem.max_memory_size = Some(*mms as usize),
+			ExecutorParam::StackLogicalMax(slm) => stack_limit.logical_max = *slm,
+			ExecutorParam::StackNativeMax(snm) => stack_limit.native_stack_max = *snm,
+			ExecutorParam::InstantiationStrategy(is) => match is {
+				ExecInstantiationStrategy::PoolingCoW =>
+					sem.instantiation_strategy = InstantiationStrategy::PoolingCopyOnWrite,
+				ExecInstantiationStrategy::RecreateCoW =>
+					sem.instantiation_strategy = InstantiationStrategy::RecreateInstanceCopyOnWrite,
+				ExecInstantiationStrategy::Pooling =>
+					sem.instantiation_strategy = InstantiationStrategy::Pooling,
+				ExecInstantiationStrategy::Recreate =>
+					sem.instantiation_strategy = InstantiationStrategy::RecreateInstance,
```

Comment:
nit: Can do something like this (haven't checked if compiles):

```suggestion
			ExecutorParam::InstantiationStrategy(is) => sem.instantiation_strategy = match is {
				ExecInstantiationStrategy::PoolingCoW => InstantiationStrategy::PoolingCopyOnWrite,
				ExecInstantiationStrategy::RecreateCoW => InstantiationStrategy::RecreateInstanceCopyOnWrite,
				ExecInstantiationStrategy::Pooling => InstantiationStrategy::Pooling,
				ExecInstantiationStrategy::Recreate => InstantiationStrategy::RecreateInstance,
```

Reply:
Yes, you're right, it definitely looks better.